# AQI Prediction Model using Python

- PM2.5 PM10
- NO, NO2
- NH3 - Ammonia
- CO
- SO2
- O3
- Benzene, toluene, Xylene

In [1]:
pip install numpy pandas matplotlib seaborn scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from warnings import filterwarnings
filterwarnings('ignore')

In [4]:
df=pd.read_csv('air quality data.csv')
df.head()

City        Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad  2015-01-01    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad  2015-01-02    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad  2015-01-03    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad  2015-01-04    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad  2015-01-05    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  133.36     0.00     0.02    0.00  NaN        NaN  
1   34.06     3.68     5.50    3.77  NaN        NaN  
2   30.70     6.80    16.40    2.25  NaN        NaN  
3   36.08     4.43    10.14    1.00  NaN        NaN  
4   39.31     7.01    18.89    2.78  NaN        NaN

In [5]:
df.shape

(29531, 16)

In [6]:
df.info

<bound method DataFrame.info of                 City        Date  PM2.5   PM10     NO    NO2    NOx    NH3  \
0          Ahmedabad  2015-01-01    NaN    NaN   0.92  18.22  17.15    NaN   
1          Ahmedabad  2015-01-02    NaN    NaN   0.97  15.69  16.46    NaN   
2          Ahmedabad  2015-01-03    NaN    NaN  17.40  19.30  29.70    NaN   
3          Ahmedabad  2015-01-04    NaN    NaN   1.70  18.48  17.97    NaN   
4          Ahmedabad  2015-01-05    NaN    NaN  22.10  21.42  37.76    NaN   
...              ...         ...    ...    ...    ...    ...    ...    ...   
29526  Visakhapatnam  2020-06-27  15.02  50.94   7.68  25.06  19.54  12.47   
29527  Visakhapatnam  2020-06-28  24.38  74.09   3.42  26.06  16.53  11.99   
29528  Visakhapatnam  2020-06-29  22.91  65.73   3.45  29.53  18.33  10.71   
29529  Visakhapatnam  2020-06-30  16.64  49.97   4.05  29.26  18.80  10.03   
29530  Visakhapatnam  2020-07-01  15.00  66.00   0.40  26.85  14.05   5.20   

          CO    SO2      O3  Be

In [15]:
df.duplicated().sum()

0

In [16]:
df.isnull().sum()

City              0
Date              0
PM2.5           678
PM10           7086
NO              387
NO2             391
NOx            1857
NH3            6536
CO              445
SO2             605
O3              807
Benzene        3535
Toluene        5826
Xylene        15372
AQI               0
AQI_Bucket        0
dtype: int64

In [17]:
df.dropna(subset=['AQI'],inplace=True)

In [18]:
df.isnull().sum().sort_values(ascending=False)

Xylene        15372
PM10           7086
NH3            6536
Toluene        5826
Benzene        3535
NOx            1857
O3              807
PM2.5           678
SO2             605
CO              445
NO2             391
NO              387
City              0
Date              0
AQI               0
AQI_Bucket        0
dtype: int64

In [19]:
df.shape


(24850, 16)

In [20]:
df.describe().T

count        mean         std    min      25%      50%       75%  \
PM2.5    24172.0   67.476613   63.075398   0.04  29.0000   48.785   80.9250   
PM10     17764.0  118.454435   89.487976   0.03  56.7775   96.180  150.1825   
NO       24463.0   17.622421   22.421138   0.03   5.6600    9.910   20.0300   
NO2      24459.0   28.978391   24.627054   0.01  11.9400   22.100   38.2400   
NOx      22993.0   32.289012   30.712855   0.00  13.1100   23.680   40.1700   
NH3      18314.0   23.848366   25.875981   0.01   8.9600   16.310   30.3600   
CO       24405.0    2.345267    7.075208   0.00   0.5900    0.930    1.4800   
SO2      24245.0   14.362933   17.428693   0.01   5.7300    9.220   15.1400   
O3       24043.0   34.912885   21.724525   0.01  19.2500   31.250   46.0800   
Benzene  21315.0    3.458668   16.036020   0.00   0.2300    1.290    3.3400   
Toluene  19024.0    9.525714   20.881085   0.00   1.0275    3.575   10.1800   
Xylene    9478.0    3.588683    6.754324   0.00   0.3900    1.420    4.1200   
AQI      24850.0  166.463581  140.696585  13.00  81.0000  118.000  208.0000   

             max  
PM2.5     914.94  
PM10      917.08  
NO        390.68  
NO2       362.21  
NOx       378.24  
NH3       352.89  
CO        175.81  
SO2       186.08  
O3        257.73  
Benzene   455.03  
Toluene   454.85  
Xylene    170.37  
AQI      2049.00

In [21]:
null_values_percentage=(df.isnull().sum()/df.isnull().count()*100).sort_values(ascending=False)
null_values_percentage

Xylene        61.859155
PM10          28.515091
NH3           26.301811
Toluene       23.444668
Benzene       14.225352
NOx            7.472837
O3             3.247485
PM2.5          2.728370
SO2            2.434608
CO             1.790744
NO2            1.573441
NO             1.557344
City           0.000000
Date           0.000000
AQI            0.000000
AQI_Bucket     0.000000
dtype: float64

## Key Consideration
 - Xylene has the highest percentage of missing values - 61.88%
 - PM10 and NH3 28 - 26%